### PR TITLE
Batch Checkouts When Upgrading Overrides

### PR DESCRIPTION
**Pull request by NickGerleman**
react-native-platform-override currently upgrades overrides sequentially, for each one checking out an old version of RN from the scratch Git repo along with a new one. This is costly because checkouts are expensive, and we do multiple per upgrade.

This change makes it so that upgrades are triggered concurrently, and adds a mechanism to GitReactFileRespository to batch together requests of the same React Native version. This dramatically reduces checkout frequency.

An informal benchmark was done integrating 2.5 months of updates:
- Previous time: **1m20s**
- New time: **33s**

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5495)

### CODE REVIEW

File: packages/react-native-platform-override/src/BatchingQueue.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      this.keyedQueues.get(key).push(async () => {
+        try {
+          resolve(await action());
+        } catch (ex) {
+          reject(ex);
+        }
+      });
+
+      if (!this.currentKey) {
+        this.currentKey = key;
```
Comment (dannyvv) [inline packages/react-native-platform-override/src/BatchingQueue.ts:30]: >key [](start = 26, length = 3)

nit: perhaps a check for if key has a value, else the logic goes wrong..

File: packages/react-native-platform-override/src/BatchingQueue.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+
+/**
+ * Executes actions, attempting to group by a given key
+ */
+export default class BatchingQueue<TKey> {
+  private keyedQueues: Map<TKey, Array<() => Promise<void>>> = new Map();
+  private currentKey?: TKey;
+
+  enqueue<T>(key: TKey, action: () => Promise<T>): Promise<T> {
```
Comment (dannyvv) [inline packages/react-native-platform-override/src/BatchingQueue.ts:15]: >key [](start = 13, length = 3)

nit: perhaps a check for if key has a value, else the logic goes wrong..

File: packages/react-native-platform-override/src/BatchingQueue.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    const currentQueue = this.keyedQueues.get(this.currentKey!);
+
+    while (currentQueue.length > 0) {
+      await currentQueue.shift()!();
+    }
+
+    this.keyedQueues.delete(this.currentKey!);
+    this.currentKey = undefined;
+
+    if (this.keyedQueues.size > 0) {
```
Comment (dannyvv) [inline packages/react-native-platform-override/src/BatchingQueue.ts:46]: >this [](start = 8, length = 4)

nit: Perhaps add a comment that says something like: finding they next key with the largest queue to process next.

File: packages/react-native-platform-override/src/BatchingQueue.ts (hunk 2 above)
Comment (NickGerleman) [reply packages/react-native-platform-override/src/BatchingQueue.ts:15]: We have strictNullChecks enabled, so the key can't be undefined or null unless we explicitly allow it.

File: packages/react-native-platform-override/src/BatchingQueue.ts (hunk 3 above)
Comment (NickGerleman) [reply packages/react-native-platform-override/src/BatchingQueue.ts:46]: Can add this

**Pull request merged**